core/divelist.c

```diff
@@ -214,21 +231,20 @@ static double calculate_cns_dive(const struct dive *dive)
 		t = sample->time.seconds - psample->time.seconds;
 		if (sample->setpoint.mbar) {
 			po2 = sample->setpoint.mbar;
-		} else {
+		} else {	// Find FiO2 as well as PO2 at mean depth of dive segment:
 			int o2 = active_o2(dive, dc, sample->time);
-			po2 = lrint(o2 * depth_to_atm(sample->depth.mm, dive));
+			po2 = lrint(o2 * depth_to_atm((int)round((sample->depth.mm + psample->depth.mm) / 2.0), dive));
 		}
-		/* CNS don't increse when below 500 matm */
-		if (po2 < 500)
+		/* Don't increase CNS when below 500 matm */
+		if (po2 <= 500)
 			continue;
-		/* Find what table-row we should calculate % for */
-		for (j = 1; j < sizeof(cns_table) / (sizeof(int) * 3); j++)
+		/* Find the table-row for calculating the maximum exposure at this PO2 */
+		for (j = 1; j < sizeof(cns_table) / (sizeof(int) * 5); j++)
 			if (po2 > cns_table[j][0])
 				break;
-		j--;
```

Comment:
Did you leave out this j-- on purpose? I don't know which direction your slope (to the left or to the right) is actually going to work.

Reply:
j-- dropped intentionally